kaiax/reward/impl/getter.go

```diff
@@ -0,0 +1,395 @@
+
+// GetDeferredReward returns the rewards to be applied at the end of the block.
+// Under non-deferred mode, transaction fees are ignored.
+func (r *RewardModule) GetDeferredReward(header *types.Header, txs []*types.Transaction, receipts []*types.Receipt) (*reward.RewardSpec, error) {
+	config, err := reward.NewRewardConfig(r.ChainConfig, r.GovModule, header)
+	if err != nil {
+		return nil, err
+	}
+	totalFee, err := getTotalFee(config, header, txs, receipts)
+	if err != nil {
+		return nil, err
+	}
+
+	return r.getDeferredReward(config, header, totalFee)
+}
+
+func (r *RewardModule) getDeferredReward(config *reward.RewardConfig, header *types.Header, totalFee *big.Int) (*reward.RewardSpec, error) {
+	if config.IsSimple {
+		return getDeferredRewardSimple(config, totalFee)
+	} else {
+		si, err := r.StakingModule.GetStakingInfo(header.Number.Uint64())
+		if err != nil {
+			return nil, err
+		}
+		return getDeferredRewardFull(config, totalFee, si)
+	}
+}
+
+// getTotalFee calculates the total transaction fees in the block.
+func getTotalFee(config *reward.RewardConfig, header *types.Header, txs []*types.Transaction, receipts []*types.Receipt) (*big.Int, error) {
+	if config.Rules.IsKaia {
+		// sum { tx[i].gasUsed * tx[i].effectiveGasPrice }
+		// = block.gasUsed * block.baseFeePerGas + sum { tx[i].gasUsed * tx[i].effectiveGasTip }
+		if len(txs) != len(receipts) {
+			return nil, reward.ErrTxReceiptsLenMismatch
+		}
+		totalFee := new(big.Int).Mul(big.NewInt(int64(header.GasUsed)), header.BaseFee)
+		for i, tx := range txs {
+			tip := new(big.Int).Mul(big.NewInt(int64(receipts[i].GasUsed)), tx.EffectiveGasTip(header.BaseFee))
+			totalFee = totalFee.Add(totalFee, tip)
+		}
+		return totalFee, nil
+	} else if config.Rules.IsMagma {
+		// Optimized to block.gasUsed * block.baseFeePerGas
+		return new(big.Int).Mul(big.NewInt(int64(header.GasUsed)), header.BaseFee), nil
+	} else {
+		// Optimized to block.gasUsed * governance.unitprice
+		return new(big.Int).Mul(big.NewInt(int64(header.GasUsed)), config.UnitPrice), nil
+	}
+}
+
+// getDeferredRewardSimple is for Simple policy.
+func getDeferredRewardSimple(config *reward.RewardConfig, totalFee *big.Int) (*reward.RewardSpec, error) {
+	spec := reward.NewRewardSpec()
+	minted := new(big.Int).Set(config.MintingAmount)
+
+	// Non-deferred mode
+	if !config.DeferredTxFee {
+		var proposer *big.Int
+		if config.Rules.IsMagma {
+			// In non-deferred mode, no fees to distribute here at the end of block processing.
+			// Just distribute the minting reward to the proposer and stop.
+			proposer = new(big.Int).Set(minted)
+			totalFee = big.NewInt(0)
+		} else {
+			// But Simple policy had a bug where transaction fees were distributed to the proposer here at the end of block processing
+			// despite configured to non-deferred mode. To keep the backward compatibility, the buggy behavior retains until Magma.
+			proposer = new(big.Int).Add(minted, totalFee)
+		}
+		spec.Minted = new(big.Int).Set(minted)
```

Comment:
```suggestion
		spec.Minted = minted
```

Reply:
I'm afraid `spec.Minted = minted` copies the pointer which potentially cause a hard to catch bug.